### ISSUE TITLE
How do I generate the .apk file of a mobile application from mooshine?

### ISSUE THREAD

**Issue by lmarevaloa** [prominic/Moonshine-IDE#454]
Hello friends, I am creating an AIR mobile application with mooshine 1.15.0, but how can I generate the .apk of my app.

**Comment by rat-moonshine** [comment]
There is no in-build feature available to generate apk/ipa in Moonshine yet.

You can use the Ant build script that supplied with the project when you create a new mobile project in Moonshine. The Ant build script can be found at:

`$projectPath/build/build.xml`

The Ant build script can work as a simple starting point for your project. You may need to extend the script if requires per your project dependencies. The supplied build script is targeted to generate apk only using self-signed certificate, but you can change the certificate details as well.

**Comment by lmarevaloa** [comment]
ok, but what command is used to compile script build.xml..

**Comment by piotrzarzycki21** [comment]
@lmarevaloa you have two options, assuming you have installed [Apache Ant](https://ant.apache.org/):
1. You can open command console (CMD - Win, Temrinal - Mac) in folder of your project `$projectPath/build/build.xml` and type `ant` - In that case you need to have environment variable `FLEX_HOME` setup 
2. You can go in Moonshine to menu `Project` -> `Build with Apache Ant`

**Comment by lmarevaloa** [comment]
I run it with cmd but I open the build.xml in an excel sheet, and the second option executes it from moonshine but the following comes out:

Moonshine IDE 1.16.0
: Source code is under Apache License, Version 2.0
: https://github.com/prominic/Moonshine-IDE
: Uses as3abc (LGPL), as3swf (MIT), fzip (ZLIB), asblocks (Apache License 2.0), NativeApplicationUpdater (LGPL)
: Running on Adobe AIR 32.0.0.89
: Ant build Running
: The system cannot find the path specified.
: The system cannot find the path specified.
: FSCH exit code: 1

**Comment by lmarevaloa** [comment]
When I use the sdk apache royale 0.9.4 FP11.7AIR14.0 (JS, SWF) I have the following error:

Compiling app
: Sending to mxmlc: mxmlc -load-config+=obj/appConfig.xml -static-link-runtime-shared-libraries=true -accessible=true +configname=airmobile -debug=true -o bin-debug/app.swf
: Error: Could not find or load main class flex2.tools.Fcsh

I must switch to apache flex 4.9.1 FP21.0 AIR21.0 to run the mobile AIR app...IN MOONSHINE

**Comment by piotrzarzycki21** [comment]
> When I use the sdk apache royale 0.9.4 FP11.7AIR14.0 (JS, SWF) I have the following error:
> 
> Compiling app
> : Sending to mxmlc: mxmlc -load-config+=obj/appConfig.xml -static-link-runtime-shared-libraries=true -accessible=true +configname=airmobile -debug=true -o bin-debug/app.swf
> : Error: Could not find or load main class flex2.tools.Fcsh
> 
> I must switch to apache flex 4.9.1 FP21.0 AIR21.0 to run the mobile AIR app...IN MOONSHINE

I completely do not understand what happened here. Apache Royale is an SDK where you can build application to SWF and JS. You cannot build Flex application trough that SDK. As for the first issue I'm trying to download Flex SDK 4.9.1 and will try to test with this. Is there any reason why are you not trying to use Flex SDK 4.16.1 ?

**Comment by rat-moonshine** [comment]
@lmarevaloa , given the simplicity of the provided Ant build script file with Moonshine mobile project, it reads following value from your system environment variable, as @piotrzarzycki21 also mentioned:
- FLEX_HOME

Please, make sure the SDK mentioned in `FLEX_HOME` env.variable is same as of your compiling SDK to the mobile project in Moonshine. There is way to inject custom FLEX_HOME path to the Ant build script but that require necessary changes in the Ant build script file.

You will also require to have Apache Ant SDK setup in your system to run the Ant build script. If you are running the Ant build script through Moonshine, then make sure you have Ant SDK setup in Moonshine settings. To do that, open `Moonshine -> Settings -> Ant Build Setup`, and provide the path to And SDK:

![image](https://user-images.githubusercontent.com/24370870/52687666-fe449800-2f78-11e9-9daa-7c13af5f4417.png)

**Comment by piotrzarzycki21** [comment]
> You will also require to have Apache Ant SDK setup in your system to run the Ant build script. If you are running the Ant build script through Moonshine, then make sure you have Ant SDK setup in Moonshine settings. To do that, open `Moonshine -> Settings -> Ant Build Setup`, and provide the path to And SDK:

It would prompt him immediately if he wouldn't have setup Ant SDK, so this is definitely not the case. 

@lmarevaloa please show us screenshot that you have setup Flex SDK either in menu **File** -> **Settings** -> **Default SDK** or right click on project **Settings** -> **Build options**.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine you are right! I'm sorry. I just reproduced the issue! @lmarevaloa you should have setup your ant path like that: 

![ant](https://user-images.githubusercontent.com/24554795/52688222-cbd57380-2f55-11e9-8fa2-67c988e0d85f.png)

**Comment by piotrzarzycki21** [comment]
I successfully build Flex Mobile app with Apache Flex 4.9.1 (Adobe Air 21.0), so once you setup proper path to ant you should be able to build app.

**Comment by rat-moonshine** [comment]
> It would prompt him immediately if he wouldn't have setup Ant SDK, so this is definitely not the case.
> I just reproduced the issue! @lmarevaloa you should have setup your ant path like that:

Does that mean the immediate popup you mentioned did not able to store Ant path effectively, but when going by `Moonshine -> Settings` ?

**Comment by piotrzarzycki21** [comment]
> Does that mean the immediate popup you mentioned did not able to store Ant path effectively, but when going by `Moonshine -> Settings` ?

It stores path correctly, but issue is reproducible when you add there totally wrong path, for example to C:/ or whatever which do not contains ANT.

**Comment by lmarevaloa** [comment]
> > When I use the sdk apache royale 0.9.4 FP11.7AIR14.0 (JS, SWF) I have the following error:
> > Compiling app
> > : Sending to mxmlc: mxmlc -load-config+=obj/appConfig.xml -static-link-runtime-shared-libraries=true -accessible=true +configname=airmobile -debug=true -o bin-debug/app.swf
> > : Error: Could not find or load main class flex2.tools.Fcsh
> > I must switch to apache flex 4.9.1 FP21.0 AIR21.0 to run the mobile AIR app...IN MOONSHINE
> 
> I completely do not understand what happened here. Apache Royale is an SDK where you can build application to SWF and JS. You cannot build Flex application trough that SDK. As for the first issue I'm trying to download Flex SDK 4.9.1 and will try to test with this. Is there any reason why are you not trying to use Flex SDK 4.16.1 ?


because I get the following error:

Compiling app
: Sending to mxmlc: mxmlc -load-config+=obj/appConfig.xml -accessible=true -static-link-runtime-shared-libraries=true +configname=airmobile -debug=true -o bin-debug/app.swf
: Error: Unable to access jarfile C:\Program Files (x86)\Adobe\Adobe Flash Builder 4.6\sdks\4.16.1\lib\fcsh.jar


I try to add the missing files but I keep getting errors :(

**Comment by lmarevaloa** [comment]
> > Does that mean the immediate popup you mentioned did not able to store Ant path effectively, but when going by `Moonshine -> Settings` ?
> 
> It stores path correctly, but issue is reproducible when you add there totally wrong path, for example to C:/ or whatever which do not contains ANT.

![image](https://user-images.githubusercontent.com/47543948/52718561-c25a0480-2f71-11e9-9774-875155e987e9.png)

**Comment by lmarevaloa** [comment]
Change the path in FLEX_HOME and I put sdk apache flex 4.9.1 but the following error occurs..

BUILD FAILED
: C:\Users\equipopc\Documents\app\build\build.xml:144: Java returned: 12
: 
: Total time: 5 seconds
: FSCH exit code: 12

![image](https://user-images.githubusercontent.com/47543948/52720895-807f8d00-2f76-11e9-898e-20974cb6955c.png)

Check the file but it is the route of the FLEX_HOME I do not understand. In the file build.xml there is more than one variable $ {ADT_PATH} but the error I get in this line

**Comment by piotrzarzycki21** [comment]
Ok summarize everything your steps are:
1. You are launching Moonshine and have there in menu **File** -> **Settings**:
A. Ant Build Setup -> Path to ANT folder
B. SDK and java is setup:
![java_sdk](https://user-images.githubusercontent.com/24554795/52728534-c0b32f80-2fb7-11e9-8852-9332797d6164.png)
2. You are creating new **Flex Mobile Project (iOS, Android)** project either using options on Home screen or in menu.
3. Once you create project you are choosing **Project** -> **Build with Apache Ant**

In the results of above steps you are getting above issues ?

**Comment by piotrzarzycki21** [comment]
The very last question is - How did you download your SDK ? Did you use [Apache Flex Installer](https://flex.apache.org/installer.html) ?

**Comment by piotrzarzycki21** [comment]
I'm not sure what is happen, but I would love to see how it looks like on your sight that you cannot even build Hello World. - It's working for me without the problem. I can offer you meeting where you could share your screen and we could go trough each step. Let me know if you are interested - we can schedule something suits your time zone.

**Comment by lmarevaloa** [comment]
> Ok summarize everything your steps are:
> 
> 1. You are launching Moonshine and have there in menu **File** -> **Settings**:
>    A. Ant Build Setup -> Path to ANT folder
>    B. SDK and java is setup:
>    ![java_sdk](https://user-images.githubusercontent.com/24554795/52728534-c0b32f80-2fb7-11e9-8852-9332797d6164.png)
> 2. You are creating new **Flex Mobile Project (iOS, Android)** project either using options on Home screen or in menu.
> 3. Once you create project you are choosing **Project** -> **Build with Apache Ant**
> 
> In the results of above steps you are getting above issues ?

I tell you that I have configured Apache flex SDK 4.9.1 for the "Hello World" app to work like this one in the image. But I run with Project-> Build & Run.

![image](https://user-images.githubusercontent.com/47543948/52747617-42a05a00-2fb2-11e9-9641-83fb37681748.png)


If I run with Project-> Build with Apache Ant I get the following error:

Total time: 7 seconds
: 
: BUILD FAILED
: C:\Users\equipopc\Documents\app\build\build.xml:144: Java returned: 12
: 
: Total time: 7 seconds

**Comment by lmarevaloa** [comment]
> The very last question is - How did you download your SDK ? Did you use [Apache Flex Installer](https://flex.apache.org/installer.html) ?

Download it from the official website http://flex.apache.org/installer.html

**Comment by lmarevaloa** [comment]
> I'm not sure what is happen, but I would love to see how it looks like on your sight that you cannot even build Hello World. - It's working for me without the problem. I can offer you meeting where you could share your screen and we could go trough each step. Let me know if you are interested - we can schedule something suits your time zone.

Yes friend, tell me the day and the time I am from Ecuador and here is 5:18 pm. If you can before or after I tell you.

**Comment by piotrzarzycki21** [comment]
> If I run with Project-> Build with Apache Ant I get the following error:
> 
> Total time: 7 seconds
> :
> : BUILD FAILED
> : C:\Users\equipopc\Documents\app\build\build.xml:144: Java returned: 12
> :
> : Total time: 7 seconds

Could you go to your `-app.xml`  file and check whether you have in the first line this one?
`<application xmlns="http://ns.adobe.com/air/application/21.0">`

If you use newest Air for example 32 with Flex SDK 4.16.1 you have to have there:
`<application xmlns="http://ns.adobe.com/air/application/32.0">`

If you didn't have setup 21 there, please do so and try to rebuild by ANT. Let me know what's the results.

**Comment by rat-moonshine** [comment]
> Could you go to your `-app.xml` file and check whether you have in the first line this one?

It looks like if SDK version mismatch, the build script returns error appropriately, though:

```
: compileAPKProject:
:      [java] C:\Users\devsena\Desktop\NewFlexMobileProject\build\build-debug\NewFlexMobileProject-app.xml: error 102: Invalid namespace http://ns.adobe.com/air/application/30.0
: 
: BUILD FAILED
: C:\Users\devsena\Desktop\NewFlexMobileProject\build\build.xml:144: Java returned: 12
: 
: Total time: 7 seconds
: 
: BUILD FAILED
```

**Comment by lmarevaloa** [comment]
> > If I run with Project-> Build with Apache Ant I get the following error:
> > Total time: 7 seconds
> > :
> > : BUILD FAILED
> > : C:\Users\equipopc\Documents\app\build\build.xml:144: Java returned: 12
> > :
> > : Total time: 7 seconds
> 
> Could you go to your `-app.xml` file and check whether you have in the first line this one?
> `<application xmlns="http://ns.adobe.com/air/application/21.0">`
> 
> If you use newest Air for example 32 with Flex SDK 4.16.1 you have to have there:
> `<application xmlns="http://ns.adobe.com/air/application/32.0">`
> 
> If you didn't have setup 21 there, please do so and try to rebuild by ANT. Let me know what's the results.


Make the change -app.xml 11.0 to 21.0 and if it worked. Thanks friend

**Comment by lmarevaloa** [comment]
> > Could you go to your `-app.xml` file and check whether you have in the first line this one?
> 
> It looks like if SDK version mismatch, the build script returns error appropriately, though:
> 
> ```
> : compileAPKProject:
> :      [java] C:\Users\devsena\Desktop\NewFlexMobileProject\build\build-debug\NewFlexMobileProject-app.xml: error 102: Invalid namespace http://ns.adobe.com/air/application/30.0
> : 
> : BUILD FAILED
> : C:\Users\devsena\Desktop\NewFlexMobileProject\build\build.xml:144: Java returned: 12
> : 
> : Total time: 7 seconds
> : 
> : BUILD FAILED
> ```
> One thing is, the compilation may complete if mentioned version is lower than the using version, but fail if mentioned version is higher than the using version.

I need to run my app with a more current version of apache sdk You know where I can download the full sdk 4.6.1 because I get the following error when I try it with that package:

: Compiling app
: Sending to mxmlc: mxmlc -load-config+=obj/appConfig.xml -accessible=true -static-link-runtime-shared-libraries=true +configname=airmobile -debug=true -o bin-debug/app.swf
: Error: Unable to access jarfile C:\apache096\royale-asjs\lib\fcsh.jar

**Comment by piotrzarzycki21** [comment]
You can download it using Apache Flex installer - did you try it? Did work?

**Comment by lmarevaloa** [comment]
> You can download it using Apache Flex installer - did you try it? Did it work?

I am testing it but the following error comes out. I put the AIR_HOME route but the error follows.

 Compiling app
: Sending to mxmlc: mxmlc -load-config+=obj/appConfig.xml -accessible=true -static-link-runtime-shared-libraries=true +configname=airmobile -debug=true -o bin-debug/app.swf
: Apache Flex Compiler SHell (fcsh)
: Version 4.16.1 build 20171115
: Copyright 2017 The Apache Software Foundation.
: 
: (fcsh) 
: FSCH target: 1
: fcsh: Assigned 1 as the compile target id
: Loading configuration file C:\apacheflexsdk4.6.1air32\frameworks\airmobile-config.xml
: 
: Loading configuration file C:\Users\equipopc\Documents\app\obj\appConfig.xml
: 
: (fcsh) 
: C:\apacheflexsdk4.6.1air32\frameworks\airmobile-config.xml(87): Error: unable to open '{airHome}/frameworks/libs/air/servicemonitor.swc'
: 
:       </library-path>
:

**Comment by piotrzarzycki21** [comment]
@lmarevaloa here is my thoughts.
1) Maybe you have some problem with permissions cause your SDK is on drive C. Move it to Documents or whatever different location
2) In order to make sure that you really have properly downloaded SDK let's download it using different method.

**Steps:**
1) Download Apache Flex SDK 4.16.1 [binaries](http://www.apache.org/dyn/closer.lua/flex/4.16.1/binaries/apache-flex-sdk-4.16.1-bin.zip)
2) Unzip it in a place where you would like to store your sdk - mabe choose different location than disk C.
3) Open Windows Command Line (CMD) and go to that folder
4) Type in folder following command:
`ant -f installer.xml -Dair.sdk.version=32.0 -Dflash.sdk.version=32.0`
You will be prompted by license stuff, type y everywhere and wait once it download. 
5) Point downloaded SDK in Moonshine and make sure that you have in your -app.xml file 
<application xmlns="http://ns.adobe.com/air/application/32.0">

One additional think is that maybe you are affected by issue which I have fixed in #447, so to make sure that it is not the case download [Development version of Moonshine](https://moonshine-ide.com/downloads/releases/dev/windows/). This file is being refreshed whenever we push some fixes etc.

Once you do that - create new mobile project. The old one may be affected by issue from #447.

**Comment by lmarevaloa** [comment]
> @lmarevaloa here is my thoughts.
> 
> 1. Maybe you have some problem with permissions cause your SDK is on drive C. Move it to Documents or whatever different location
> 2. In order to make sure that you really have properly downloaded SDK let's download it using different method.
> 
> **Steps:**
> 
> 1. Download Apache Flex SDK 4.16.1 [binaries](http://www.apache.org/dyn/closer.lua/flex/4.16.1/binaries/apache-flex-sdk-4.16.1-bin.zip)
> 2. Unzip it in a place where you would like to store your sdk - maybe choose different location than disk C.
> 3. Open Windows Command Line (CMD) and go to that folder
> 4. Type in folder following command:
>    `ant -f installer.xml -Dair.sdk.version=32.0 -Dflash.sdk.version=32.0`
>    You will be prompted by license stuff, type y everywhere and wait once it download.
> 5. Point downloaded SDK in Moonshine and make sure that you have in your -app.xml file
> 
> One additional think is that maybe you are affected by issue which I have fixed in #447, so to make sure that it is not the case download [Development version of Moonshine](https://moonshine-ide.com/downloads/releases/dev/windows/). This file is being refreshed whenever we push some fixes etc.
> 
> Once you do that - create new mobile project. The old one may be affected by issue from #447.


Download the update from the link you gave me https://moonshine-ide.com/downloads/releases/dev/windows/ .....and it worked :) . Thanks friend for helping me.

**Comment by piotrzarzycki21** [comment]
Cool! :)
